### PR TITLE
Improve crate selection on rustdoc search results page

### LINKED ISSUE #93240: Rustdoc new crate filter location visually unpolished
I took a few notes while looking at nightly rustc docs:

Screenshot_20220123_165956

@rustbot label T-rustdoc, C-bug, C-enhancement, A-rustdoc-ui

The sans serif vs. serif is clearly a bug. That’s looking terrible.

The location being in a place that moves horizontally is suboptimal; changing that would be an enhancement. The font size is ridiculously high (because it’s part of a “title”, I guess?). I like the thing no longer being inside of the search bar like it used to be, so I’m not suggesting any reverts, just more polishing work. Maybe this option could be placed e.g. still roughly next to the “Results for …” heading, but right-justified so it doesn’t move around when I change my search term, and also in a normal / more reasonable size. E.g. saying something like “search in \[all crates\]”.

By the way, something not indicated in the picture that I just noticed: if it says “… in \[All crates\]”, I don’t think the “All” should be capitalized anymore.

The question of whether or not it’s necessary for rustdoc to parrot my search term at all is one I don’t care all that much about. If there’s any benefit to keeping it, sure, keep it; I don’t really care.

<hr>

(Minor note, looking back at my image, the arrows are a bit short. The text about font size is in-fact referring to the font size of the dropdown list, nothing else.)

### PR DESCRIPTION
**Pull request by steffahn**
Improves some problems pointed out in #93240.

In particular the `in` being serif-font makes sense once this is in a new line, and the font-size is better so the dropdown menu isn’t this huge anymore. This also addresses my concern that the selection moves around horizontally as the search term changes.

The formatting for the `#search-settings` CSS is not necessary for the title, because `h1` has the same CSS, too, anyways.

Currently on nightly:
Screenshot_20220703_200911
Screenshot_20220703_200924

~~After this PR:~~ _(outdated, see below)_
Screenshot_20220703_201333

Screenshot_20220703_200829

---

With my latest changes added, I'd say this PR can close #93240 entirely. See https://github.com/rust-lang/rust/pull/98855#issuecomment-1176822718 below to see renderings of the current status.

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
I uploaded an online version of this change [here](https://rustdoc.crud.net/imperio/improve_rustdoc_search_results_page_crates_selection/foo/index.html?search=foo). I'm personally not a big fan of putting the "in X" part in its own line but let's see what others think about it.

cc @jsha @Manishearth

**Comment by steffahn** [comment]
The thing you uploaded only has a single crate, so the list doesn’t show up in the first place. It did however remind me that in that case, the `<br/>` probably shouldn’t appear in the HTML (even though it doesn’t appear to get rendered as a new line anyway), so after re-inspecting the logic in the JS, I’ve moved it into the `crates` variable to make it conditional like that.

**Comment by jsha** [comment]
I agree with @GuillaumeGomez that we shouldn't use a second line of vertical space. @steffahn you commented in #93240 that you are indifferent to whether the results page echoes the search query. I think that might be a better solution than using a second line; let's remove the "for `search_term`" part. Then the dropdown can be on the same line but not move around as you edit your search.

**Comment by steffahn** [comment]
Okay, keeping it in the title and removing the search term; making the `in` part of the title font (but not the selection, that looks weird in bold). I would still like to keep the selection dropdown smaller, if that’s stylistically acceptable. So how about something like this for a look? (I just manually edited this in the browser so far, not implemented it yet into the source files.)

Screenshot_20220706_192234

With a single crate, the thing would *just* say Results.

**Comment by steffahn** [comment]
On second thought, in e.g. the white theme where the dropdown is lower-contrast, keeping matching fonts seems kind-of essential actually. Compare:

Screenshot_20220706_193911

Screenshot_20220706_193924

the second looks better IMO.

**Comment by steffahn** [comment]
> theme where the dropdown is lower-contrast

Makes me reconsider perhaps also addressing the other point I made in #93240 here, the high contrast in the dark theme currently. Is there any other dropdown-selections in rustdoc that should be taken into consideration for consistency?

**Comment by steffahn** [comment]
I added an implementation of a style along those lines, i.e. as explained in my previous two comments.

Screenshot_20220707_003120
Screenshot_20220707_003155
Screenshot_20220707_003229

Small font size in dropdown:
Screenshot_20220707_003414

As you might be able to tell by looking into the CSS, I had a hard time adapting the color of that arrow to the field; I’m open for suggestions for cleaner approaches if an inline svg like this is not acceptable.

File: src/librustdoc/html/static/js/search.js [hunk 1]
```diff
@@ -1710,22 +1704,15 @@ function initSearch(rawSearchIndex) {
         let crates = "";
         const crates_list = Object.keys(rawSearchIndex);
         if (crates_list.length > 1) {
-            crates = " in <select id=\"crate-search\"><option value=\"All crates\">" +
-                "All crates</option>";
+            crates = " in <select id=\"crate-search\"><option value=\"all crates\">" +
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/js/search.js:1707]: Why this change?

Comment (steffahn) [reply src/librustdoc/html/static/js/search.js:1707]: because "Result in all crates" reads better than "Result in All crates". Or do you mean the internal representation changing from "All crates" to "all crates", too? That just felt more consistent than just changing the label, but if there's a reason against it, it can of course be kept "All crates" internally.

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:1707]: I just checked locally and I think it actually renders quite nicely locally.

Comment (steffahn) [reply src/librustdoc/html/static/js/search.js:1707]: By the way, feel free to do your magic again for uploading an online versions of this change for others to check out.

File: src/librustdoc/html/static/css/themes/ayu.css [hunk 2]
```diff
@@ -220,10 +220,22 @@ details.undocumented > summary::before {
...
+.search-input {
 	background-color: #141920;
+	border-color: #424c57;
+}
+#crate-search {
+	background-color: #0f1419;
 	/* Without the `!important`, the border-color is ignored for `<select>`... */
-	border-color: #424c57 !important;
+	border-color: #5c6773 !important;
+	background-image: url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.or\
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/css/themes/ayu.css:231]: I gave a try to a svg minifier (just in case) and the result is indeed a bit shorter mostly because of the `<svg>` attributes:

```
data:image/svg+xml;utf8,<svg width="128" height="128" enable-background="new 0 0 128 128" version="1.1" viewBox="-30 -20 176 176" xml:space="preserve" xmlns="http://www.w3.org/2000/svg"><g fill="none" stroke="%235c6773" stroke-linecap="square" stroke-miterlimit="10" stroke-width="12"><line x1="111" x2="64" y1="40.5" y2="87.499"/><line x1="64" x2="17" y1="87.499" y2="40.5"/></g></svg>
```

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/css/themes/ayu.css:231]: Just realized but please put `background-image` into `rustdoc.css` to prevent copying it into multiple places.

File: src/librustdoc/html/static/js/search.js (hunk 1 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/js/search.js:1707]: I'll after your next update. I really like this change. :)

File: src/librustdoc/html/static/css/themes/ayu.css (hunk 2 above)
Comment (steffahn) [reply src/librustdoc/html/static/css/themes/ayu.css:231]: It's a different svg depending on the theme (the color is different). That's the reason I've put it inline in the first place, since despite some searching, I couldn't find a nice way to modify the color of the existing svg from CSS. I just copied the pre-existing file `down-arrow.svg` and modified the color. I'm open to better solutions. E. g. to use `filter`, which can change color, one would somehow need to make the arrow it's own element or pseudo-element, presumably involving some helper-`div`. I don't know, perhaps I could give such a restructure a try, but I'm really not all that familiar with html and css and this stuff in the first place, so it's quite a bit of effort.

Comment (steffahn) [reply src/librustdoc/html/static/css/themes/ayu.css:231]: The status quo is particularly unclean because of the forced line breaks because the tidy check otherwise fails the CIt.

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/css/themes/ayu.css:231]: Ah I see. I had to do that recently so I can maybe help you with that. Give me a few minutes. :smile:

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/css/themes/ayu.css:231]: You can disable the tidy check on a line iirc. To be confirmed.

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/css/themes/ayu.css:231]: I see. Sorry, misunderstood what you meant previously. Then yes, you need to use a pseudo element.

**Comment by GuillaumeGomez** [comment]
Just realized that if you don't need to use `background-image` in all themes, we already have it in `rustdoc.css`.

**Comment by GuillaumeGomez** [comment]
That seems very weird to switch the "focus" from the main content to the sidebar, no?

**Comment by steffahn** [comment]
I have yet to find out why the thing has so much more visual padding in the `<select>` on Firefox than Chrome, and how to avoid this. (I prefer the way it looks on Chrome / that’s what I looked at when changing/choosing the padding in the first place.)

Chrome
Screenshot_20220709_000525

Firefox
Screenshot_20220709_000546

**Comment by steffahn** [comment]
Some screenshots of the hover effects I’ve added yesterday:
Screenshot_20220709_094014
Screenshot_20220709_094209
Screenshot_20220709_094105
Screenshot_20220709_094142
Screenshot_20220709_093903
Screenshot_20220709_094236

For light and ayu, colors of outline (both with and without focus, respectively) are the same as the settings and help button; the border color without focus is also the same as the outline of the settings *menu* on all themes, so that’s how I chose the outline color for the dark theme; the blue hover color of the dark theme is the same as the hover color for the radio buttons and the activate switches in the settings menu, admitted there are a lot of slightly different shades of blue, i.e. the searchbar highlight above (if it’s focused) is a different blue, and the active tab below (“In Names”/“In Parameters”/“In Return Types”) has yet-another shade of blue.

File: src/librustdoc/html/static/css/themes/ayu.css (hunk 2 above)
Comment (steffahn) [reply src/librustdoc/html/static/css/themes/ayu.css]: In the latest commit, I use a filter like that now, with a `:after` pseudo-element and a `div`. I hope everything works (or at least is not in a badly broken state) in all browsers we want to support; I would appreciate if you were able to create an online version again so I (and perhaps others) can test this on more devices.

**Comment by GuillaumeGomez** [comment]
I uploaded the docs generated with this PR [here](https://rustdoc.crud.net/imperio/improve_rustdoc_search_results_page_crates_selection/foo/index.html?search=test).

**Comment by steffahn** [comment]
@GuillaumeGomez

I’m going to resolve the merge conflict by improving upon #99086 by keeping the title in a single line
Screenshot_20220710_231838

---

Next up, I’ll also squash the commits and rebase; the current history is mostly me fighting with CI or completely re-designing everything, anyway.

**Comment by steffahn** [comment]
@GuillaumeGomez, I have no idea whatsoever how these `.goml` tests work in detail, and how I need to adjust / re-write the test to test for the new improved behavior I’ve implemented. So I feel like I’m unable to make this pass CI on my own. I would appreciate some help, e.g. explanation of what these tests even mean / what *exactly* they test, line-by-line. (I’m referring to the ones introduced in #99086.) Or alternatively, maybe you could simply write an appropriate updated test for me? Either posted here, or perhaps I’d disable the test and you can create a new one in a separate PR?

(It doesn’t help that last time I tried, I had trouble getting these tests to run on my machine in the first place, and running them with CI takes about half an hour until I get feedback.)

**Comment by steffahn** [comment]
I’ve disabled the abovementioned failing (still rather new) tests for now; if it’s okay to have them only come back in a future PR, then this PR should be ready to be reviewed and merged from my end, provided CI passes now 🚀

File: src/librustdoc/html/static/css/rustdoc.css [hunk 3]
```diff
@@ -934,14 +949,35 @@ table,
 	-webkit-appearance: none;
 	/* Removes default arrow from firefox */
 	text-indent: 0.01px;
-	text-overflow: "";
+}
+/* cancel stylistic differences in padding
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/css/rustdoc.css:953]: When does it happen? We don't use `appearance: none` ourselves after all.

File: src/librustdoc/html/static/css/themes/dark.css [hunk 4]
```diff
@@ -196,11 +196,26 @@ details.undocumented > summary::before {
...
-	border-color: #f0f0f0 !important;
+	border-color: #d2d2d2;
+}
+#crate-search-div::after {
+	/* match border-color; uses https://codepen.io/sosuke/pen/Pjoqqp */
+	filter: invert(94%) sepia(0%) saturate(721%) hue-rotate(255deg) brightness(90%) contrast(90%);
+}
+#crate-search:hover, #crate-search:focus {
+	border-color: #2196f3 !important;
+	border-color: #008dfd !important;
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/css/themes/dark.css:215]: Something weird is going on here. :)

File: src/test/rustdoc-gui/search-result-display.goml [hunk 5]
```diff
@@ -15,26 +15,28 @@ assert-css: (".search-results div.desc", {"width": "570px"})
...
 size: (900, 900)
 
-// First we check the current width and position.
-assert-css: ("#crate-search", {"width": "222px"})
-compare-elements-position-near: (
-    "#crate-search",
-    "#search-settings .search-results-title",
-    {"y": 5},
-)
+// FIXME: Fix and re-enable these tests!
```
Comment (GuillaumeGomez) [inline src/test/rustdoc-gui/search-result-display.goml:18]: So about the `goml` format. It's something I wrote to make it easy to write/read. You can see more information about it (with the full list of commands) [here](https://github.com/GuillaumeGomez/browser-UI-test/blob/master/goml-script.md#command-list).

File: src/librustdoc/html/static/css/rustdoc.css (hunk 3 above)
Comment (steffahn) [reply src/librustdoc/html/static/css/rustdoc.css:953]: I think we effectively are, kind-of… I guess… at least with `appearance: none` it looks the same, IDK about backwards-compatibility with older versions of Firefox where it might’ve been different..?

In any case, what this is supposed to say is that Firefox has some (empirically, 4px) padding in an "unstyled" <select> (i.e. when the native arrow is removed) that Chrome doesn't have. I don't have any other browsers myself ATM to test / compare.

Admitted, given that `appearance: none` is **not** literally what’s used here, the comment could be clarified.

File: src/librustdoc/html/static/css/themes/dark.css (hunk 4 above)
Comment (steffahn) [reply src/librustdoc/html/static/css/themes/dark.css:215]: Looks like some copy-paste error, oops…

I’ll double-check which color is correct.

File: src/librustdoc/html/static/css/rustdoc.css (hunk 3 above)
Comment (steffahn) [reply src/librustdoc/html/static/css/rustdoc.css]: I changed the wording a bit now.

I’m not sure about the pros and cons of using the “actual” `appearance: none` property, I haven’t researched too much what distinguishes this property from those `-moz-appearance` / `-webkit-appearance` ones; consequently I kept everything as-is with regards to these properties, just to be safe, e.g. since I haven’t tested other browsers, etc…

**Comment by steffahn** [comment]
> We can take a look at the `goml` tests at the end if you're still stuck. Not really the hard part normally. ;)

Well for one thing, I’d need to get those tests running locally in the first place, otherwise it’s impossible to work with.

It just gives me a completely cryptic error message
```
Running 66 rustdoc-gui (8 concurrently) ...
ERROR: puppeteer failed when trying to create a new page. Please try again with `--no-sandbox`
```

I’m also noticing this warning on the `npm install browser-ui-test`:
```
npm WARN deprecated puppeteer@3.3.0: Version no longer supported. Upgrade to @latest
```

I have no idea if that’s related to the problem and if yes how to fix it.

**Comment by steffahn** [comment]
W.r.t. rewriting/adapting the tests, they currently (judging by the comments) seem to work with the implicit assumption that the `<select>` lives in its own line and just test for its size. What’s *actually* interesting is presumably that its right-hand-side border is positioned within-bounds of, say, the `search` section, I have no idea how one would write this.

I also don’t know what this `compare-elements-position-near`-stuff means or what the error message talking about `"#search-settings .search-results-title"` means, in particular what the `"#search-settings .search-results-title"` syntax itself means. Looks like two CSS selectors separated by space; IDK how those compose.

**Comment by steffahn** [comment]
TBH, I don’t have much more time I can spend on this today. I would highly appreciate if I didn’t have to learn this whole testing infrastructure now, too. This PR is already so much more that I initially wanted to do. Learning new HTML and CSS fundamentals I didn’t know before is enough commitment for my liking, I don’t feel like going too deep into this whole testing framework, too, now.

**Comment by GuillaumeGomez** [comment]
I can do this part then.

**Comment by GuillaumeGomez** [comment]
I took the commits from here, fixed the conflicts and added the missing GUI tests in https://github.com/rust-lang/rust/pull/100374.

**Comment by steffahn** [comment]
Looks like at some point Firefox must have changed, at least my Firefox browser no longer shows the differences that used to be compensated with the
```css
/* cancel stylistic differences in padding in firefox
for "appearance: none"-style (or equivalent) <select>s */
@-moz-document url-prefix() {
	#crate-search {
		padding-left: 0px; /* == 4px - 4px */
		padding-right: 19px; /* == 23px - 4px */
	}
}
```
part, hence that can (and should) probably be removed. I’ll need to look into this more, and create an appropriate new issue and PR if this is really the case. (Makes me wonder whether there are good ways to do something like “display a website for a bunch of different versions of Firefox” in order to confirm that there really was a change at some version throughout the last year, and I’m not the one that changed a factor.

**Comment by GuillaumeGomez** [comment]
I was wondering the same thing a few days ago. Don't hesitate to send a PR in any case!

As for checking on different firefox versions, I remember that websites allow it, but don't remember the name. ^^'

**Comment by steffahn** [comment]
I didn’t find any websites easily. I’ll just ignore the urge to conclusively “prove” something in Firefox has changed, and just go with the observation that both on my PC and phone, the current Firefox seems to be displaying it differently than intended and removing the special case workaround CSS code seems like the way to fix it. So I’ll just open a PR tomorrow. :-)

**Comment by GuillaumeGomez** [comment]
Perfect, thanks!
